File: src/maze.c
//
// Created by efy on 12/09/25.
//

#include "vec2i.h"
#include "maze.h"

#include <math.h>
#include <stdlib.h>
#include <string.h>

Point* grid_geti(const Grid* grid, const uint8_t x, const uint8_t y) {
    const uint8_t i = x + y * HEIGHT;

    if (i < 0 || i > SIZE) {
        return NULL;
    }

    return &grid->points[i];
}

static uint8_t ceili(const double cx) {
    return (int)ceil(cx);
}

Point* grid_getd(const Grid* grid, const double x, const double y) {
    return grid_geti(grid, ceili(x), ceili(y));
}

Point* grid_get_relative(const Grid* grid, const Point* point, const Direction direction) {
    switch (direction) {
        case NORTH:
            return grid_geti(grid, point->pos.x, point->pos.y - WIDTH);
        case WEST:
            return grid_geti(grid, point->pos.x - 1, point->pos.y);
        case EAST:
            return grid_geti(grid, point->pos.x + 1, point->pos.y);
        case SOUTH:
            return grid_geti(grid, point->pos.x, point->pos.y + WIDTH);
        default:
            return NULL;
    }
}

// Amount of cells in the middle of the maze, the goal.
#define GOAL_CELL_COUNT 4

// The coordinates of the mid-cells in the maze.
static const Vec2i GOAL_CELLS[GOAL_CELL_COUNT] = {
    {WIDTH / 2 - 1, HEIGHT / 2 - 1},
    {WIDTH / 2, HEIGHT / 2 - 1},
    {WIDTH / 2 - 1, HEIGHT / 2},
    {WIDTH / 2, HEIGHT / 2},
};

/**
 * Returns the smallest value of two numbers.
 * @param x Value one.
 * @param y Value two.
 * @return The smallest value. If x equals y, returns x.
 */
static uint8_t min(const uint8_t x, const uint8_t y) {
    if (x <= y) {
        return x;
    }
    return y;
}

/**
 * Returns the grid distance to the nearest goal.
 * @param x The x coordinate.
 * @param y The y coordinate.
 * @return The grid distance to the nearest goal.
 */
static uint8_t grid_distance(const uint8_t x, const uint8_t y) {
    uint8_t umin = UINT8_MAX;

    for (int i = 0; i < GOAL_CELL_COUNT; ++i) {
        const Vec2i* vec2i = &GOAL_CELLS[i];
        const uint8_t distance = abs(x - vec2i->x) + abs(y - vec2i->y);

        umin = min(umin, distance);
    }

    return umin;
}

void grid_init(const Grid* grid, Point* history, uint8_t* history_size) {
    for (uint8_t x = 0; x < WIDTH; ++x) {
        for (uint8_t y = 0; y < HEIGHT; ++y) {
            Point* point = grid_geti(grid, x, y);

            point->pos = (Vec2i){x, y};
            point->distance = grid_distance(x, y);
            for (int j = 0; j < DIRECTION_LENGTH; ++j) {
                point->wall[j] = NULL;
            }
        }
    }

    history[0] = grid->points[0];
    *history_size = *history_size + 1;
}

File: src/maze.rs
use crate::pathfinder::Target;
use crate::vec::Vecu;
use crate::{MAZE_HEIGHT_U8, MAZE_SIZE, MAZE_WIDTH_U8};
use core::slice::Iter;

/// Represents the maze
pub struct Maze {
    segments: [Segment; MAZE_SIZE],
}

/// Calculates the distance to the specified point
fn maze_calc_distance(x: u8, y: u8, cx: i8, cy: i8) -> u8 {
    (i8::abs((x as i8) - cx) + i8::abs((y as i8) - cy)) as u8
}

/// Converts the position to the index for `segments`.
fn xy_to_index(x: u8, y: u8) -> usize {
    (x + y * MAZE_WIDTH_U8) as usize
}

/// Converts the position to the index for `segments`.
fn pos_to_index(pos: Vecu) -> usize {
    xy_to_index(pos.x, pos.y)
}

/// Creates a new Maze with the provided walls.
fn with_walls_fn<T>(target: Target, walls: T) -> Maze
where
    T: Fn(u8, u8) -> [bool; 4],
{
    let mut points = [Segment::new(); MAZE_SIZE];

    let distances;

    match target {
        Target::Center => {
            distances = [
                |x, y| maze_calc_distance(x, y, 8, 8),
                |x, y| maze_calc_distance(x, y, 7, 8),
                |x, y| maze_calc_distance(x, y, 8, 7),
                |x, y| maze_calc_distance(x, y, 7, 7),
            ];
        }
        Target::Origin => {
            distances = [
                |x, y| maze_calc_distance(x, y, 0, 0),
                |_x, _y| u8::MAX,
                |_x, _y| u8::MAX,
                |_x, _y| u8::MAX,
            ];
        }
    }

    for x in 0..MAZE_WIDTH_U8 {
        for y in 0..MAZE_HEIGHT_U8 {
            points[xy_to_index(x, y)] = Segment {
                pos: Vecu { x, y },
                distance: *distances.map(|f| f(x, y)).iter().min().unwrap(),
                walls: walls(x, y),
            };
        }
    }

    Maze { segments: points }
}

impl Maze {

    /// Creates a new maze centered around the center.
    pub fn new() -> Self {
        with_walls_fn(Target::Center, |_x, _y| [false, false, false, false])
    }

    /// Creates a new maze with the specified target and the existing maze to use walls from.
    pub fn with_walls(target: Target, maze: Maze) -> Self {
        with_walls_fn(target, |x, y| maze.segment(x, y).walls)
    }

    /// Returns the segment at `x, y`.
    pub fn segment(&self, x: u8, y: u8) -> Segment {
        self.segments[xy_to_index(x, y)]
    }

    /// Returns the segment at `x, y`.
    pub fn segment_vec(&self, pos: Vecu) -> Segment {
        self.segment(pos.x, pos.y)
    }

    /// Adds a wall at the specified direction.
    fn _add_wall(&mut self, x: u8, y: u8, direction: Relative) {
        let i = xy_to_index(x, y);
        let mut existing = self.segments[i];
        existing.walls[direction as usize] = true;
        self.segments[i] = existing;
    }

    /// Updates the walls of the segment at `x, y` to the specified array.
    pub fn update_walls(&mut self, x: u8, y: u8, walls: [bool; 4]) {
        let i = xy_to_index(x, y);
        let mut existing = self.segments[i];

        for (i, val) in existing.walls.iter().enumerate() {
            assert!(
                (!(*val) && !walls[i]) || (!(*val) && walls[i]) || (*val && walls[i]),
                "Walls can only be updated from false to true"
            )
        }

        existing.walls = walls;
        self.segments[i] = existing;

        // update neighbouring segments' walls
        for (j, dir) in Relative::iter().enumerate() {
            if !walls[j] {
                continue;
            }

            let relative = existing.relative(self, dir);
            if relative.is_none() {
                continue;
            }

            let mut relative = relative.unwrap();
            relative.walls[dir.opposite() as usize] = true;
            self.segments[pos_to_index(relative.pos())] = relative;
        }
    }

    /// Updates the distance of the segment at `x, y` to the specified value.
    pub fn update_distance(&mut self, x: u8, y: u8, distance: u8) {
        let i = (x + y * MAZE_WIDTH_U8) as usize;
        let existing = self.segments[i];

        self.segments[i] = Segment {
            pos: existing.pos,
            distance,
            walls: existing.walls,
        };
    }
}

impl Default for Maze {
    fn default() -> Self {
        Self::new()
    }
}

impl core::fmt::Debug for Maze {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        for y in 0..MAZE_HEIGHT_U8 {
            for x in 0..MAZE_WIDTH_U8 {
                let distance = self.segment(x, y).distance;
                if distance < 10 {
                    write!(f, "{:?}  ", distance)?;
                } else {
                    write!(f, "{:?} ", distance)?;
                }
            }
            writeln!(f)?;
        }
        writeln!(f)?;
        Ok(())
    }
}

/// The relative direction.
#[derive(PartialEq, Debug)]
pub enum Relative {
    North,
    East,
    South,
    West,
}

impl Relative {
    /// Returns an iterator over all relative directions.
    pub fn iter() -> Iter<'static, Relative> {
        static DIRECTIONS: [Relative; 4] = [
            Relative::North,
            Relative::East,
            Relative::South,
            Relative::West,
        ];
        DIRECTIONS.iter()
    }

    /// Returns the direction opposite to this direction.
    pub fn opposite(&self) -> Relative {
        match self {
            Relative::North => Relative::South,
            Relative::East => Relative::West,
            Relative::South => Relative::North,
            Relative::West => Relative::East,
        }
    }
}

/// Represents a point on the grid.
#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Segment {
    pos: Vecu,
    pub distance: u8,
    pub walls: [bool; 4],
}

impl Segment {
    /// Creates a new default Segment.
    pub fn new() -> Self {
        Segment {
            pos: Vecu::new(),
            distance: u8::MAX,
            walls: [false, false, false, false],
        }
    }

    /// Creates a new default Segment.
    pub fn with_pos(pos: Vecu) -> Self {
        Segment {
            pos,
            distance: u8::MAX,
            walls: [false, false, false, false],
        }
    }

    /// Whether this segment is a dead end.
    /// A dead end is defined as a segment with 3 walls and 1 entrance.
    pub fn is_dead_end(&self) -> bool {
        self.walls.iter().filter(|it| **it).count() == 3
    }

    /// Whether this segment is straight.
    /// A segment is straight when there are 2 walls and 1 exit and 1 entrance.
    pub fn is_straight(&self) -> bool {
        self.walls.iter().filter(|it| **it).count() == 2
    }

    /// Returns the position of this segment.
    pub fn pos(&self) -> Vecu {
        self.pos
    }

    /// Returns the segment relative to this one by direction `relative`.
    ///
    /// ### Arguments
    ///
    /// - `maze` - A maze ref.
    /// - `relative` - The direction.
    pub fn relative(&self, maze: &Maze, relative: &Relative) -> Option<Segment> {
        let (x, y) = (self.pos.x, self.pos.y);

        match *relative {
            Relative::North if y > 0 => Some(maze.segment(x, y - 1)),
            Relative::South if y + 1 < MAZE_HEIGHT_U8 => Some(maze.segment(x, y + 1)),
            Relative::West if x > 0 => Some(maze.segment(x - 1, y)),
            Relative::East if x + 1 < MAZE_WIDTH_U8 => Some(maze.segment(x + 1, y)),
            _ => None,
        }
    }
}

impl Default for Segment {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use crate::maze::Maze;
    use std::panic::{catch_unwind, AssertUnwindSafe};

    #[test]
    fn test_wall_update_panics() {
        let mut maze = Maze::new();
        maze.update_walls(0, 0, [false, true, false, false]);
        assert!(catch_unwind(AssertUnwindSafe(|| maze.update_walls(
            0,
            0,
            [false, false, false, false]
        )))
        .is_err())
    }

    #[test]
    fn test_wall_update() {
        let mut maze = Maze::new();
        maze.update_walls(0, 0, [false, true, false, false]);

        assert_eq!([false, true, false, false], maze.segment(0, 0).walls);
        assert_eq!([false, false, false, true], maze.segment(1, 0).walls);

        let mut maze = Maze::new();
        maze.update_walls(0, 0, [false, false, true, false]);

        assert_eq!([false, false, true, false], maze.segment(0, 0).walls);
        assert_eq!([true, false, false, false], maze.segment(0, 1).walls);

        let mut maze = Maze::new();
        maze.update_walls(1, 0, [false, false, false, true]);

        assert_eq!([false, false, false, true], maze.segment(1, 0).walls);
        assert_eq!([false, true, false, false], maze.segment(0, 0).walls);
    }
}
